src/arena.c

```diff
@@ -1101,13 +1103,21 @@ arena_cache_bin_fill_small(tsdn_t *tsdn, arena_t *arena,
 	malloc_mutex_lock(tsdn, &bin->lock);
 	arena_bin_flush_batch_after_lock(tsdn, arena, bin, binind, &batch_flush_state);
 
-	while (filled < nfill) {
+	while (filled < nfill_max) {
+		if (filled >= nfill_min) {
+			/* Drop the last fill if already filled nfill_min. */
+			assert(filled != 0);
+			break;
+		}
+		assert(filled < nfill_min);
 		/* Try batch-fill from slabcur first. */
 		edata_t *slabcur = bin->slabcur;
 		if (slabcur != NULL && edata_nfree_get(slabcur) > 0) {
-			unsigned tofill = nfill - filled;
-			unsigned nfree = edata_nfree_get(slabcur);
-			unsigned cnt = tofill < nfree ? tofill : nfree;
+			/* Use up the free slots if the total filled <= nfill_max. */
+			unsigned cnt = edata_nfree_get(slabcur);
+			if (cnt + filled > nfill_max) {
+				cnt = nfill_min - filled;
```

Comment:
Did you meant to have `cnt = nfill_max - filled`? The `nfill_min - filled` would imply we always fill to min?